ecosystem/sep-0023.md

```diff
@@ -0,0 +1,140 @@
+## Simple Summary
+
+Strkey is an ASCII format for representing Stellar account IDs and
+addresses.  This document introduces augmented strkey to represent
+account IDs that are multiplexed across users or separate transactions
+and hence must be used in conjunction with a particular memo ID.
+
+## Motivation
+
+A common pattern in the Stellar ecosystem is for services to share a
+single Stellar account ID across many users or independent
+transactions, relying on the memo ID to disambiguate incoming
+payments.  Unfortunately, experience shows that people frequently
+forget to include the memo ID, resulting in either lost funds or
+onerous support calls.  Hence, we need a way to specify an address and
+memo together in a way that is difficult for end-users to misuse.
+
+## Abstract
+
+A new strkey format is introduced for _multiplexed_ addresses that can
+only be used as a destination address and only in conjunction with a
+particular memo ID.
+
+## Specification
+
+To multiplex a Stellar account over multiple users, the account must
+be used in conjunction with a particular memo ID.  Because the memo
+field is per-transaction rather than per-operation, this places
+restrictions on the use of such multiplexing.  In particular:
+
+* A multiplexed address can occur only as the `destination` field of a
+  `PAYMENT`, `PATH_PAYMENT`, or `ACCOUNT_MERGE` operation.
+
+* At most one multiplexed address can appear in a given transaction.
+
+### Native representation
+
+Because multiplexed addresses cannot be used everyplace a normal
+account can be, implementations must represent them using a distinct
+type.  Since this representation depends on the language, we give a
+suggested implementation in XDR:
+
+~~~ {.c}
+struct MultiplexedAccountID {
+  uint64 *memoID;
+  AccountID accountID;
+};
+~~~
+
+If `memoID` is `NULL`, the structure should be considered equivalent
+to a simple `AccountID`.  Otherwise, it is a multiplexed `AccountID`
+subject to the restrictions above, and must have a `MEMO_ID`.
+
+### Encoding
+
+Strkey provides a compact ASCII format for ED25519 public keys,
+ED25519 private keys (also known as seeds), pre-authorized transaction
+hashes, and hash-x signers (which provide signing authority upon
+revelation of a SHA-256 preimage). We add a fifth type representing
+multiplexed account IDs named by ED25519 keys.  Each of these types
+has a corresponding version byte, which determines the first character
+of the strkey encoding:
+
+| Key type                | Version byte | First char | Muxed |
+| ----------------------- | ------------ | ---------- | ----- |
+| STRKEY\_PUBKEY\_ED25519 | 6 << 3       | G          | no    |
+| STRKEY\_PUBKEY\_ED25519 | 12 << 3      | M          | yes   |
+| STRKEY\_SEED\_ED25519   | 18 << 3      | S          | no    |
+| STRKEY\_PRE\_AUTH\_TX   | 19 << 3      | T          | no    |
+| STRKEY\_HASH\_X         | 23 << 3      | X          | no    |
+
+
+The following steps transform a binary key into a strkey:
```

Comment:
The problem with this algorithm is that base32 symbols are 5 bit and (1+8+32+2)*8 % 5 == 3. This may require devs to update existing strkey encoders/decoders code if it doesn't support padding. Maybe it's worth adding 2 extra null bytes after a version byte? It will increase strkey encoded address by 3 chars though.

Reply:
It's worst than that: when the payload is not a multiple of 5 bits, libraries have to implement a version without padding but because the RFC discourages doing that, not using padding bytes is likely not supported/buggy.
For example, I found multiple bugs in just the base32 encoders/decoders that we use in core when non aligned payloads were involved (originally, I was not using 8 bits for the version when designing "StrKey" and ran into this issue).

So if we pad the payload with 2 extra bytes, we add a total of 10 bytes of data, 56+16=72 characters which is not that different from the version that relies on proper handling of non padded non aligned.

Reply:
This sounds like a valid concern. 

What if we change the Multiplexed Version byte `0x60` (12 << 3) to be multiple bytes `0x65 0x17 0x3f`? 

This adds the two bytes required to avoid padding and introduces a prefix of `MULT` instead of just `M`.

Reply:
First off, if you are padding, RFC4648 requires `=` signs if the input is not a multiple of 5 *bytes*, not 5 bits.  However, for crypto agility we are going to need the ability to have arbitrary length public keys.  That means any padding should be in the Strkey specification, not the underlying XDR, as otherwise we'll keep having this problem.

I personally think that we should just require the variant without padding.  This stuff isn't that hard to implement and people shouldn't be relying on third-party libraries anyway, unless those libraries are highly trusted (like built into your language runtime).  Someone who controls these padding libraries can steal funds.

That said, if we really want to pad, then we should pad with `=` as specified in the RFC.  We should definitely not invent a new padding, and even less violate layering and start placing restrictions on XDR key formats just so that the can be base-32 encoded a certain way.